cpp/tests/iterator/iterator_test.cu

```diff
@@ -0,0 +1,393 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Proof the concept of iterator driven aggregations to reuse the logic
+
+   The concepts:
+   1. computes the aggregation by given iterators
+   2. computes by using cub and thrust with same function parameters
+   3. accepts nulls and group_by with same function parameters
+
+    CUB reduction:  https://nvlabs.github.io/cub/structcub_1_1_device_reduce.html#aa4adabeb841b852a7a5ecf4f99a2daeb
+    Thrust reduction: https://thrust.github.io/doc/group__reductions.html#ga43eea9a000f912716189687306884fc7
+
+    Thrust iterators: https://thrust.github.io/doc/group__fancyiterator.html
+*/
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <bitset>
+#include <cstdint>
+#include <iostream>
+#include <numeric>
+#include <iterator>
+
+#include <utilities/cudf_utils.h> // need for CUDA_HOST_DEVICE_CALLABLE
+#include <utilities/device_atomics.cuh> // need for device operators.
+
+#include <tests/utilities/column_wrapper.cuh>
+#include <tests/utilities/cudf_test_fixtures.h>
+
+#include <cub/device/device_reduce.cuh>
+#include <thrust/device_vector.h>
+#include <thrust/transform.h>
+
+#include <thrust/tuple.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/iterator_adaptor.h>
+
+
+// --------------------------------------------------------------------------------------------------------
+// structs for column_input_iterator
+template<typename T>
+struct ColumnDataNonNull{
+    const T *data;
+
+    __device__ __host__
+    ColumnDataNonNull(T *_data)
+    : data(_data){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        return data[id];
+    };
+};
+
+template<typename T>
+struct ColumnData{
+    const T *data;
+    const gdf_valid_type *valid;
+    T identity;
+
+    __device__ __host__
+    ColumnData(T *_data, gdf_valid_type *_valid, T _identity)
+    : data(_data), valid(_valid), identity(_identity){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        return (gdf_is_valid(valid, id))? data[id] : identity;
+    };
+};
+
+template<typename T>
+struct ColumnDataSquare : public ColumnData<T>
+{
+    ColumnDataSquare(T *_data, gdf_valid_type *_valid, T _identity)
+    : ColumnData<T>(_data, _valid, _identity){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        T val = ColumnData<T>::at(id);
+        return (val * val);
+    };
+};
+
+// TBD. ColumnDataSquareNonNull
+
+
+// column_input_iterator
+template<typename T_output, typename T_input, typename Iterator=thrust::counting_iterator<gdf_index_type> >
```

Comment:
So making sure I understand this...

`T_output` is passed into `column_input_iterator` so that you can specify the type of `super_t::reference`?

Reply:
Yes, the 6th (= last -1) template parameter of  the parent `thrust::iterator_adaptor` is the `super_t::reference`.

https://github.com/rapidsai/cudf/blob/c971adf6dbf87e337a9d0db8ed535259f38b3e78/cpp/tests/iterator/iterator_test.cu#L108-L111